ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -1258,14 +1333,35 @@ checkTime cfgEnv intEnv =
          KnownIntersectionState blk
       -> (LedgerState blk -> Maybe a)
       -> WithEarlyExit m (Intersects blk a)
-    readLedgerState kis prj = castM $ readLedgerStateHelper kis prj
+    readLedgerState kis prj = do
+        let blockingCall ::
+                 (   KnownIntersectionState blk
+                  -> KnownIntersectionState blk
+                 )
+              -> m (WithEarlyExit m (Intersects blk a))
+            blockingCall upd = readLedgerStateHelper (upd kis) prj retry
+        -- The nesting of these calls is so that we 'stopIdling' before
+        -- blocking.
+        let nonBlockingCall :: m (WithEarlyExit m (Intersects blk a))
+            nonBlockingCall =
+                readLedgerStateHelper
+                    kis
+                    prj
+                    (return $ castM $ do
+                        stopIdling
+                        blockingCall (\x -> x{ theyAreIdling = False })
+                    )
+
+        castM $ if theyAreIdling kis then nonBlockingCall else blockingCall id
```

Comment:
After trying to re-convince myself, I've decided that the approach in this PR is a dead-end.

I've opened https://github.com/IntersectMBO/ouroboros-network/pull/4791 instead. I'm about to push a commit that uses that upstream PR instead of all the `theyAreIdling` logic in this PR. Unfortunately, GHC hangs when I compile that commit locally. Here's hoping the various CI servers handle that gracefully 🤞 (the memory is not growing unbounded, at least).